ENH: stats.Binomial: add binomial distribution with new infrastructure

mdhaber opened the pull request:
#### Reference issue
Closes gh-17891[^1]
Addresses gh-18564
Addresses gh-11026[^2]

#### What does this implement/fix?
This adds the binomial distribution in the new infrastructure as `scipy.stats.Binomial`. To support discrete RVs, I renamed `ContinuousDistribution` to `_UnivariateDistribution` and created subclasses `ContinuousDistribution` and `DiscreteDistribution`. The `DiscreteDistribution` subclass doesn't require too much modification:

- Wherever "quadrature" is done, use `nsum` on the PMF instead of `tanhsinh` on the PDF. ([comment](https://github.com/scipy/scipy/pull/22312#discussion_r1912521674), [comment]( https://github.com/scipy/scipy/pull/22312#discussion_r1912522530))
- When using "quadrature" to evaluate a distribution function, replace the argument with `floor(x)` (CDF) or `floor(x+1)` (CCDF).  ([comment](https://github.com/scipy/scipy/pull/22312#discussion_r1912524834))
- When inverting a distribution function numerically, set `xatol=1` and use the floor of the right end of the resulting bracket (and handle edge cases). ([comment](https://github.com/scipy/scipy/pull/22312#discussion_r1912522132), [comment](https://github.com/scipy/scipy/pull/22312#discussion_r1912525124))
- When finding the mode with numerical optimization, set a coarse[^3] `xatol` and check the `floor` and `ceil` of the result to determine which is the mode[^3]. ([comment](https://github.com/scipy/scipy/pull/22312#discussion_r1912522746))
- Adjust the `_set_invalid_nan` decorator so PMF at non-integral args is 0, and make sure the distribution functions at the left end of the support are NOT replaced (as they can be for continuous distributions). ([comment](https://github.com/scipy/scipy/pull/22312#discussion_r1912519492), [comment](https://github.com/scipy/scipy/pull/22312#discussion_r1912519877))

It also updates `make_distribution` for use with discrete distributions. All transformations are explicitly disabled (for now); thoughts on follow-ups below.

#### Additional information
The PR is in a draft state.
- Formulas for `Binomial` should be added where possible. (Relying on the infrastructure to compute everything from the PMF has been useful during informal testing.)
- We need to add tests. It would be good for someone else to take a close look at the testing strategy to see if we missed anything for `ContinuousDistribution`. @steppi, would you be willing to write them, perhaps adding formulas for the binomial distribution at the same time (so the tests can easily compare the formulas against the generic implementations)?
- I need to check which transformations work with `DiscreteDistribution`. Transformations should probably only accept instances of `ContinuousDistribution` by default (and raise otherwise). (*See https://github.com/scipy/scipy/pull/22312#issuecomment-2588114565.*)
- The "correct" behavior of two-arg distribution functions is not obvious when `y < x`. Currently, the input validation of the public functions swaps the arguments and flips the sign, which may not be what we want. Other behavior that might be more natural include (e.g. for `cdf`):
  - `cdf(y) - cdf(x) + pmf(x)`
  - 0
  - NaN

One shortcoming I'd rather address in a follow-up: currently, methods that work with numerical inversion are slow to recognize the special case when there is no valid bracket within the support. They get the right result (e.g. `Binomial(n=10, p=0.5).icdf(1e-10)` returns `0`) and it's not too slow when the distribution has compact support, but it can be slow when the support is infinite. It would be good to recognize that the support is not a valid bracket and special-case accordingly.

[^1]: In the sense that the *new* infrastructure replaces non-integral values of integral shape parameters with NaN, and consequently, methods will produce NaNs. This is consistent with the policy established in gh-21050: an error due to one bad element of an array should not cause the rest to fail.
[^2]: In the old infrastructure, `ppf(0)` and `isf(1)` are defined to be the left end of the support minus 1 (e.g. `stats.binom(n=10, p=0.5).ppf(0)  # 0`). Since `cdf` and `ccdf` are not invertible for discrete distributions, there are multiple ways `icdf` and `iccdf` can be defined, but I don't think the old definition was as useful as the new one, which returns the left end of the support. This is what R discrete distributions (e.g. `qbinom(0, 10, 0.5)`) and Mathematica (`Quantile[BinomialDistribution[10, 0.5], 0]`) seem to do. See the revised documentation of CDF/CCDF functions.
[^3]: See inline comment.
[^4]: Assumes unimodality, but that's an underlying assumption of the bracket finder / minimizer. The [documentation of `mode`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.Normal.mode.html#scipy.stats.Normal.mode) addresses this.

mdhaber committed:
Begin to draft discrete distribution

mdhaber committed:
ENH: stats.DiscreteDistribution: basic functionality for several methods

mdhaber committed:
ENH: DiscreteDistribution: make inverse functions *almost* work; need to fix ccdf/logccdf edge cases

mdhaber committed:
MAINT: fix _draw, some edge cases

mdhaber committed:
DOC: stats.DiscreteDistribution: fix examples

[skip ci]

mdhaber committed:
ENH: _UnivariateDistribution.plot: adapt to discrete distributions

mdhaber committed:
MAINT: DiscreteDistribution.i_: fix inverse method edge cases

mdhaber committed:
ENH: _ProbabilityDistribution: update considering discrete RVs

mdhaber committed:
ENH: DiscreteDistribution: adjust edge cases

mdhaber committed:
MAINT: DiscreteDistribution.mode: return integer mode

mdhaber committed:
MAINT: stats.DiscreteDistribution: avoid NaNs/warnings in entropy and nsum

mdhaber commented on scipy/integrate/_tanhsinh.py:
```diff
@@ -1308,7 +1308,7 @@ def _integral_bound(f, a, b, step, args, constants, xp):
     else:
         tol = tol + rtol*lb.integral
-    i_skip = lb.status < 0  # avoid unnecessary f_evals if integral is divergent
+    i_skip = lb.status == -3  # avoid unnecessary f_evals if integral is divergent
```
For infinite sums, `nsum` estimate the magnitude of the sum with an integral before doing anything else. This allows us to turn the relative tolerance into an absolute tolerance, which is needed when figuring out how many terms of the sum to evaluate.

This line was supposed to provide an early-exit for divergent integrals, but it (incorrectly) assumes that the only reason for failure status is that the integral is divergent. When testing with the discrete Laplace distribution, I saw that this integral failed to meet the desired tolerance when integrating across the centerline of the distribution. This is reasonable, since the function is not smooth. Consequently, the result of the whole operation was NaN, but if we remove the early-exit, we get an accurate result. The fix is to make the condition stricter - only fail here if the result of the integral is non-finite. It's a bit slower in the exceptional case, but I think that's a good trade.

mdhaber commented on scipy/stats/_distribution_infrastructure.py:
```diff
@@ -366,44 +368,6 @@ def contains(self, item, parameter_values=None):
         return in_left & in_right
 
-
-class _RealDomain(_SimpleDomain):
```
This is mostly just *moved* to below, and I moved the `draw` method to `_SimpleDomain`.

mdhaber commented on scipy/stats/_distribution_infrastructure.py:
```diff
@@ -448,8 +415,8 @@ def draw(self, n, type_, min, max, squeezed_base_shape, rng=None):
             # make this work for infinite bounds
-            z = min_nn - rng.uniform(size=shape)
-            zr = max_nn + rng.uniform(size=shape)
+            z = min_nn - uniform(1, 5, size=shape)
+            zr = max_nn + uniform(1, 5, size=shape)
```
Maybe we should comment that `5` is arbitrary; and finite number is fine.

mdhaber commented on scipy/stats/_distribution_infrastructure.py:
```diff
@@ -866,6 +897,9 @@ def filtered(self, x, *args, **kwargs):
         x = np.asarray(x)
         dtype = self._dtype
         shape = self._shape
+        discrete = isinstance(self, DiscreteDistribution)
+        keep_low_endpoint = discrete and method_name in {'_cdf1', '_logcdf1',
+                                                         '_ccdf1', '_logccdf1'}
```
Part of change 5. For continuous distributions, `cdf(0)` is replaced with `0` and `ccdf(0)` is replaced with `1`. We don't want to do this for discrete distributions.

mdhaber commented on scipy/stats/_distribution_infrastructure.py:
```diff
@@ -911,6 +945,13 @@ def filtered(self, x, *args, **kwargs):
             any_endpoint = (mask_endpoint if mask_endpoint.shape == ()
                             else np.any(mask_endpoint))
 
+        # Check for non-integral arguments to PMF method
```
Part of change 5: "Adjust the `_set_invalid_nan` decorator so PMF at non-integral args is 0".

We might want to floor the argument of other distribution methods so public methods would only call private methods with integer arguments?

mdhaber commented on scipy/stats/_distribution_infrastructure.py:
```diff
@@ -1307,9 +1355,9 @@ def _generate_example(dist_family):
-    p = 0.32
-    x = round(X.icdf(p), 2)
-    y = round(X.icdf(2 * p), 2)
+    prob = 0.32
```
I thought this was interfering with the shape parameter `p`, but in retrospect I don't think this is necessary.

mdhaber commented on scipy/stats/_distribution_infrastructure.py:
```diff
@@ -1344,19 +1392,22 @@ def _generate_example(dist_family):
     True
+    """
 
+    if issubclass(dist_family, ContinuousDistribution):
```
The first example here fails for discrete distributions because the the inverse distribution functions are very sensitive to numerical error of the input (e.g. subtractive cancellation of `1 - X.ccdf(x)`) near the edges.

The second example is two-arg CDF, which is not implemented for discrete distributions right now.

mdhaber commented on scipy/stats/_distribution_infrastructure.py:
```diff
@@ -1372,22 +1423,39 @@ def _generate_example(dist_family):
     True
+    """
+
+    if issubclass(dist_family, ContinuousDistribution):
```
`DiscreteDistribution` doesn't implement `logentropy` yet.

mdhaber commented on scipy/stats/_distribution_infrastructure.py:
```diff
@@ -1985,12 +2053,18 @@ def _quadrature(self, integrand, limits=None, args=None,
+        else:
+            res = np.asarray(nsum(f, a, b, args=args, log=log, tolerances=dict(rtol=rtol)).sum)
+            zero = -np.inf if log else 0
+            res[a > b] = zero  # nsum should probably return 0 instead of nan
```
Should we change `nsum`?

mdhaber commented on scipy/stats/_distribution_infrastructure.py:
```diff
@@ -1985,12 +2053,18 @@ def _quadrature(self, integrand, limits=None, args=None,
         # For now, we ignore the status, but I want to return the error
         # estimate - see question 5 at the top.
-        return res.integral
+        if isinstance(self, ContinuousDistribution):
```
Change 1: use `nsum` instead of `tanhsinh` for discrete distributions.

mdhaber commented on scipy/stats/_distribution_infrastructure.py:
```diff
@@ -1999,7 +2073,10 @@ def _solve_bounded(self, f, p, *, bounds=None, params=None):
         if not p.size:
             # might need to figure out result type based on p
-            return np.empty(p.shape, dtype=self._dtype)
+            res = _RichResult()
+            empty = np.empty(p.shape, dtype=self._dtype)
+            res.xl, res.x, res.xr = empty, empty, empty
+            res.fl, res.fr = empty, empty
```
Kind of clumsy/annoying, but it allows the methods that call `_solve_bounded` to treat this as any other case. I figure it's better to special-case here rather than everwhere that calls this function.

mdhaber commented on scipy/stats/_distribution_infrastructure.py:
```diff
@@ -2034,8 +2111,10 @@ def f2(x, _p, **kwargs):  # named `_p` to avoid conflict with shape `p`
         xrtol = None if _isnull(self.tol) else self.tol
-        return _chandrupatla(f3, a=res.xl, b=res.xr, args=args, xrtol=xrtol).x
+        xatol = None if xatol is None else xatol
+        return _chandrupatla(f3, a=res.xl, b=res.xr, args=args, xrtol=xrtol, xatol=xatol)
```
Change 3: when inverting a discrete distribution function numerically, set `xatol=1` to avoid unnecessary iteration. The integer root is the only integer within the bracket. (Special cases - if the bracket was not valid or if the final bracket happens to have integer endpoints - are handled in the functions that call `_solve_bounded`. The right thing to do depends on whether it's the CDF or CCDF.)

Note that this now returns the whole result object so the caller method can see the result bracket and function values. This is why we had to change the `if not p.size` case above, and in the continuous inverse methods, we now have to extract attribute `x`.

mdhaber commented on scipy/stats/_distribution_infrastructure.py:
```diff
@@ -2182,8 +2261,8 @@ def _logentropy_logexp_safe(self, **params):
         def logintegrand(x, **params):
-            logpdf = self._logpdf_dispatch(x, **params)
-            return logpdf + np.log(0j+logpdf)
+            logpxf = self._logpxf_dispatch(x, **params)
```
Part of change 1: referring to "PXF" instead of "PDF"/"PMF" does the right thing for both continuous and discrete distributions. Note that `self._quadrature` below uses `tanhsinh` for continuous and `nsum` for discrete.

mdhaber commented on scipy/stats/_distribution_infrastructure.py:
```diff
@@ -2209,9 +2288,12 @@ def _entropy_logexp(self, **params):
+            pxf = self._pxf_dispatch(x, **params)
+            logpxf = self._logpxf_dispatch(x, **params)
+            temp = np.asarray(pxf)
+            i = (pxf != 0)  # 0 * inf -> nan; should be 0
```
Surprised this hadn't come up before.

mdhaber commented on scipy/stats/_distribution_infrastructure.py:
```diff
@@ -2249,17 +2331,18 @@ def _mode_dispatch(self, method=None, **params):
         res_b = _bracket_minimum(f, m, xmin=a, xmax=b, args=args)
-        res = _chandrupatla_minimize(f, res_b.xl, res_b.xm, res_b.xr, args=args)
+        res = _chandrupatla_minimize(f, res_b.xl, res_b.xm, res_b.xr,
+                                     args=args, xatol=xatol)
```
Part of Change 4: using "coarse" `xatol`.

mdhaber commented on scipy/stats/_distribution_infrastructure.py:
```diff
@@ -2361,6 +2444,43 @@ def _pdf_formula(self, x, **params):
+    def _logpmf_dispatch(self, x, *, method=None, **params):
+        if self._overrides('_logpmf_formula'):
+            method = self._logpmf_formula
+        elif _isnull(self.tol):  # ensure that developers override _logpmf
```
Just mirroring `_pdf_dispatch` logic here. This was always a placeholder for better logic. The concern is that the PMF can underflow before the log is taken, and the hope is that annoying failures here will encourage distributions authors to override `_logpmf_formula`.

mdhaber commented on scipy/stats/_distribution_infrastructure.py:
```diff
@@ -3226,7 +3346,7 @@ def logintegrand(x, order, logcenter, **params):
     ### Convenience
 
-    def plot(self, x='x', y='pdf', *, t=('cdf', 0.0005, 0.9995), ax=None):
+    def plot(self, x='x', y=None, *, t=None, ax=None):
```
Lots of changes here to make pretty plots of discrete distribution PMFs / CDFs. This also improves the handling of distributions with finite support endpoints.

mdhaber commented on scipy/stats/_distribution_infrastructure.py:
```diff
@@ -3417,6 +3567,117 @@ def plot(self, x='x', y='pdf', *, t=('cdf', 0.0005, 0.9995), ax=None):
+            return True
+        return super()._overrides(method_name)
+
+    # These should probably check whether pmf = 0
```
Is it worth the time to check whether `np.isneginf(logpmf)`? This is typically not going to be true within the theoretical support. Also, for distributions that only define PMF, underflows would cause PDF to be zero at points that theoretically have nonzero probability density. I figure it's better for developers to override this where necessary.

steppi replied:
I had this same thought when looking through this code off GitHub, and came here to comment to find you already mentioned it.

I think the way it should go is that the pdf should be infinite for all points in the support and zero for all points out of the support. Even if the` `pmf` is zero for a point in the support, I think it's safe to act under the assumption that this is just underflow. If the probability at a point is truly zero, I think that point should be taken out of the support. Do you think that make sense?

mdhaber replied:
Yup, that's what I had in mind. I'll adjust the comment.

mdhaber replied:
How does this look?
```suggestion
    # We assume the PDF is infinite at all integers within the support.
    # Even if `_pmf_formula` is numerically zero within the support, this could easily
    # be the result of underflow, so we ignore it. If a `_logpmf_formula` is provided,
    # we *could* trust it to be -inf only where the PMF is exactly zero, but we do not.
    # Consequently, developers should override `_pdf_formula`/`_logpmf_formula` to
    # return the correct result (0 / -inf) wherever the PMF is theoretically zero within
    # the support. (Or, if there is sufficient demand, we could provide for the
    # definition of disconnected support.)
```

steppi replied:
Looks good to me.

mdhaber commented on scipy/stats/_distribution_infrastructure.py:
```diff
@@ -3417,6 +3567,117 @@ def plot(self, x='x', y='pdf', *, t=('cdf', 0.0005, 0.9995), ax=None):
+    def _logpxf_dispatch(self, x, *, method=None, **params):
+        return self._logpmf_dispatch(x, method=method, **params)
+
+    def _cdf_quadrature(self, x, **params):
+        return super()._cdf_quadrature(np.floor(x), **params)
+
+    def _logcdf_quadrature(self, x, **params):
+        return super()._logcdf_quadrature(np.floor(x), **params)
+
+    def _ccdf_quadrature(self, x, **params):
+        return super()._ccdf_quadrature(np.floor(x + 1), **params)
+
+    def _logccdf_quadrature(self, x, **params):
+        return super()._logccdf_quadrature(np.floor(x + 1), **params)
```
Part of change 2: When using "quadrature" to evaluate a distribution function, replace the argument with `floor(x)` (CDF) or `floor(x+1)` (CCDF).

mdhaber commented on scipy/stats/_distribution_infrastructure.py:
```diff
@@ -3417,6 +3567,117 @@ def plot(self, x='x', y='pdf', *, t=('cdf', 0.0005, 0.9995), ax=None):
+    def _logccdf_quadrature(self, x, **params):
+        return super()._logccdf_quadrature(np.floor(x + 1), **params)
+
+    def _icdf_inversion(self, x, **params):
+        res = self._solve_bounded(self._cdf_dispatch, x, params=params, xatol=1)
+        res = np.where(res.fl >= 0, np.floor(res.xl), np.floor(res.xr))
+        res[np.isnan(x)] = np.nan
+        return res
+
+    def _ilogcdf_inversion(self, x, **params):
+        res = self._solve_bounded(self._logcdf_dispatch, x, params=params, xatol=1)
+        res = np.where(res.fl >= 0, np.floor(res.xl), np.floor(res.xr))
+        res[np.isnan(x)] = np.nan
+        return res
+
+    def _iccdf_inversion(self, x, **params):
+        res = self._solve_bounded(self._ccdf_dispatch, x, params=params, xatol=1)
+        res = np.where(res.fl <= 0, np.floor(res.xl), np.floor(res.xr))
+        res[np.isnan(x)] = np.nan
+        return res
+
+    def _ilogccdf_inversion(self, x, **params):
+        res = self._solve_bounded(self._logccdf_dispatch, x, params=params, xatol=1)
+        res = np.where(res.fl <= 0, np.floor(res.xl), np.floor(res.xr))
+        res[np.isnan(x)] = np.nan
+        return res
```
Here's the special case handling in change 3. We might be able to eliminate it if we handle the case in which there is no valid bracket.

mdhaber commented on scipy/stats/_distribution_infrastructure.py:
```diff
@@ -3417,6 +3567,117 @@ def plot(self, x='x', y='pdf', *, t=('cdf', 0.0005, 0.9995), ax=None):
+        # convergence criterion were implemented, but the actual convergence
+        # criterion is different, and I'm not sure if it controls the total width
+        # of the bracket. `xatol=0.1` might not be safe as-implemented.
+        x = super()._mode_optimization(xatol=0.1, **params)
```
Part of change 4. `xatol=0.1` is arbitrary. Ideally we could set it to `xatol = 1` to guarantee that the minimum has been narrowed down to an interval of unit width, but see gh-22309.

mdhaber commented on scipy/stats/_probability_distribution.py:
```diff
@@ -30,8 +30,8 @@ def support(self):
-        The following table summarizes the value returned by methods
-        of ``ContinuousDistribution`` for arguments outside the support.
+        The following table summarizes the value returned by several
+        methods when the argument is outside the support.
```
Please review doc updates carefully.

mdhaber committed:
MAINT: stats.DiscreteDistribution: linting/adjustments for CI

mdhaber committed:
DOC: stats.DiscreteDistribution: fix doc issues, hopefully

[docs only]

mdhaber committed:
DOC: stats.DiscreteDistribution: additional doc fixes

[docs only]

mdhaber committed:
MAINT: stats: update transformed distributions for continuous only

mdhaber committed:
ENH: stats.make_distribution: update for DiscreteDistribution

[skip ci]

mdhaber committed:
ENH: stats.make_distribution: add support for discrete distributions

mdhaber commented on scipy/stats/tests/test_continuous.py:
```diff
@@ -1031,19 +1031,25 @@ def test_shapes(self):
                 'johnsonsb', 'kappa4', 'ksone', 'kstwo', 'kstwobign', 'powerlognorm',
-                'powernorm', 'recipinvgauss', 'studentized_range', 'vonmises_line'}
+                'powernorm', 'recipinvgauss', 'studentized_range', 'vonmises_line',
+                'betanbinom', 'zipf', 'logser', 'skellam'}  # discrete
```
`betabinom`, `zipf`, and `logser` are all slow because for the parameters tested, the median is at the left end of the support, so we run into the last shortcoming of the [top post](https://github.com/scipy/scipy/pull/22312#issue-2782644701).

mdhaber commented on scipy/stats/tests/test_continuous.py:
```diff
@@ -1031,19 +1031,25 @@ def test_shapes(self):
+            'levy_stable',            # private methods seem to require >= 1d args
+            'vonmises',               # circular distribution; shouldn't work
+            'poisson_binom',          # vector shape parameter
+            'hypergeom',              # distribution functions need interpolation
```
```python3
from scipy.special import _ufuncs as scu
M, n, N = 20, 7, 12
scu._hypergeom_pmf(3, n, N, M)  # np.float64(0.19865841073271417)
scu._hypergeom_pmf(4, n, N, M)  # np.float64(0.3575851393188855)
scu._hypergeom_pmf(3.5, n, N, M)  # np.float64(nan)
```
It would be helpful for the infinite sums if this would interpolate for non-integral arguments. We can add that, of course, but often this is something that happens naturally in the underlying implementation but is manually disabled. Any chance that's the case and we can enable it?

steppi replied:
The Boost Math implementation branches based on the size of `N`, and the large `N` implementation could be adapted to handle non-integral arguments, though it is currently constrained to unsigned values. 

https://github.com/boostorg/math/blob/a5c062542afb2037054c4e1b1be44db7b0859c38/include/boost/math/distributions/detail/hypergeometric_pdf.hpp#L467-L472

For the smaller `N` case, one coud evaluate the definition in terms of `binom`.

This is completely doable. I guess it's probably best to ask the Boost devs if they'd consider it.

mdhaber replied:
@mborland Some of the discrete distribution PMFs from Boost (e.g. binomial and negative binomial) are defined for non-integral values of the argument `k` whereas the hypergeometric distribution PMF returns NaN. There are some situations in which it is useful to approximate infinite sums involving the PDF with an integral, so if it is natural to define the hypergeometric distribution PMF for non-integral arguments, we'd appreciate it. Would you be willing/able to do that? If not, we can do our own interpolation; it would just be a bit slower. Thanks for considering it!

mborland replied:
I opened an [issue in math](https://github.com/boostorg/math/issues/1240) since this seems like a reasonable ask. I tagged you both so that you know when it's closed out.

steppi replied:
Thank you

mdhaber commented on scipy/stats/tests/test_continuous.py:
```diff
@@ -1031,19 +1031,25 @@ def test_shapes(self):
+            'vonmises',               # circular distribution; shouldn't work
+            'poisson_binom',          # vector shape parameter
+            'hypergeom',              # distribution functions need interpolation
+            'zipfian',                # distribution functions need interpolation
```
Actually I can fix this one just by adding `k = np.floor(k)` to the CDF. We could also add that to the `_dispatch` method to make sure that no `_cdf_formula` is ever called with non-integral `k`. Not sure if we want to pull the trigger on that.

mdhaber commented on scipy/stats/tests/test_continuous.py:
```diff
@@ -1031,19 +1031,25 @@ def test_shapes(self):
+            'poisson_binom',          # vector shape parameter
+            'hypergeom',              # distribution functions need interpolation
+            'zipfian',                # distribution functions need interpolation
+            'nchypergeom_fisher',     # distribution functions don't accept NaN
```
I think the arguments have to be *integers*, but we use `nan`s to when parameters are invalid in hope that functions fast-track them.
```python3
import numpy as np
from scipy import stats
M, n, N, odds = 140, 80, 60, 0.5
stats.nchypergeom_fisher._pmf(np.nan, M, n, N, odds)
# ValueError: cannot convert float NaN to integer
```
Can the wrapper just return NaN whenever any argument is a NaN (before converting to integer)?

After that, we'd face the problem that the PMF needs to interpolate... maybe doing that in the wrapper is a tall order?

mdhaber commented:
Other follow-ups that should be relatively straightforward:
- I'm not sure how useful it would be, but it doesn't look like shifting and scaling (*multiplying*) by integers would be a problem. I tested this briefly by:
  - giving `ContinuousDistribution` and `DiscreteDistribution` an attribute (e.g. `self._type`) that identifies the instance as continuous or discrete.
  - making `TransformedDistribution` a subclass of `UnivariateDistribution`, and making sure that `TransformedDistribution` copies the `_type` attribute in the initializer. 
  - Replacing some `isinstance` checks with checking the `_type` attribute.
  - (detail) Making the `p` argument of `_shift_scale_inverse_function.wrapped` positional-only to avoid name collisions with shape parameters. (This will be needed in some other places, too.)
 
  After that, a shifted/scaled `Binomial` seemed to work just fine.
  For more general shifting/scaling, some of the code assumes that the supported points are integers. Even if we generalize that, the user experience might not be great due to inherent limitations of floating point arithmetic: in general, the PMF would be nonzero and the CDF would jump only at very specific floating point values; this might be problematic if there is round-off error (either when the user calculates the arguments or we unshift/unscale them). I think there is some relevant discussion at https://github.com/scipy/scipy/pull/15932#issuecomment-1086949899 and nearby.

I'll extend this post with thoughts on `truncate` and `Mixture` soon.

I don't think I'll study other transformations in detail right now.
- `abs` and `**` (with integers on the left or right) would probably work. If they would be useful, I'll leave them to others.
- `log` and `exp` are not in the cards any time soon because the supported points would no longer be integral or even arithmetically spaced.
- I don't really want to think about order statistics in the case of discrete distributions, but this paper [discusses it](https://www.math.wm.edu/~leemis/2005informsjoc.pdf).

mdhaber committed:
DOC: stats.DiscreteDistribution: improve doc rendering

[docs only]

mdhaber committed:
ENH: DiscreteDistribution: add 2-arg cdf

mdhaber committed:
ENH: stats.DiscreteDistribution: implement logentropy

mdhaber committed:
ENH: stats.make_distribution: support zipfian distribution

mdhaber commented on scipy/stats/_new_distributions.py:
```diff
@@ -367,6 +367,34 @@ def _pdf_formula(self, x, *, a, **kwargs):
+
+    def _pmf_formula(self, x, *, n, p, **kwargs):
+        return special.binom(n, x) * p**x * (1 - p)**(n - x)
+
```
(obviously add implementations of the other methods before merge)

mdhaber committed:
Merge remote-tracking branch 'upstream/main' into DiscreteDistribution2

mdhaber committed:
MAINT: stats.DiscreteDistribution: fixup after merge

mdhaber committed:
STY: stats.DiscreteDistribution: fix lint

[lint only]

steppi committed:
MAINT: Add formulas for binomial distribution

steppi committed:
Change _SimpleDomain to _SimpleInterval to avoid ambiguity

A simply connected domain in topology is a simply connected open set,
but not all of these are open

steppi committed:
MAINT: Improve __str__ for  _IntegerInterval

steppi committed:
MAINT: Update cdf2 and friends to use probability definition

steppi committed:
TST: Update tests in response to changes in cdf2 and friends

steppi committed:
DOC: Update definition in docs for two arg ccdf

steppi committed:
MAINT: Have error message leave open future support for transforms on discrete RVs

steppi committed:
MAINT: Fix pdf of discrete distribution outside of support

steppi committed:
MAINT: Handle edge-case in discrete mode calculation

steppi committed:
TST: Update a changed warning message in test

steppi committed:
MAINT: Improvements to help tests pass

steppi committed:
MAINT: Fix edge case for pdf/pmf of discrete distribution

- non-integral x, invalid parameters

steppi committed:
TST: Skip some tests that don't make sense for discrete distributions

steppi committed:
Fix discrete cdf/logcdf for nan inputs

steppi committed:
BUG: Fix unhandled NaN input in _binom_cdf

steppi committed:
TST: Fix cdf2 tests to accomodate discrete distributions

steppi committed:
MAINT: Fix typo * instead of **

steppi committed:
MAINT: Broadcast a boolean array to correct shape

steppi committed:
MAINT: Add cdf2 safe subtraction methods to discrete dists

- dispatch was going to the parents, which isn't correct

steppi committed:
MAINT: Exclude degenerate case from tests

steppi committed:
MAINT: Fix bug, using subtraction safe when x == y

steppi committed:
MAINT: Fix reference formula for ccdf2

steppi committed:
MAINT: Fix icdf inversion and attempt to fix iccdf inversion

steppi committed:
MAINT: Fix icdf and friends inversion for real

steppi committed:
MAINT: Fix bug I introduced by changing things for debugging purposes

steppi committed:
MAINT: Introduce continuous distributions back into test_support_moments_sample

steppi committed:
TST: Add skips for rv_generic to get all tests to pass

- these seem to be due to using float internally as integer params in
discrete distributions in new infra, but using int sometimes for
discrete distributions in old infra

steppi committed:
MAINT: Remove unnecessary parens mistakenly left in

mdhaber committed:
Merge remote-tracking branch 'upstream/main' into DiscreteDistribution2

mdhaber committed:
MAINT: stats.DiscreteDistribution: replace _lazywhere with apply_where

steppi committed:
MAINT: Remove unnecessary out_of_support in pdf/logpdf formulas

steppi committed:
MAINT: Remove unnecessary order == 0 cases in moment formulas

steppi committed:
Get rid of excessively complex transform logic for x > y case

mdhaber committed:
Merge remote-tracking branch 'upstream/main' into DiscreteDistribution2

steppi committed:
MAINT: Fix typo, fillvalue instead of fill_value

steppi committed:
MAINT: Take order 0 out of list of moment formula orders

steppi committed:
MAINT: Ensure apply_where gets tuples not lists for args

steppi committed:
MAINT: Use xp_promote to convert numerical endpoints to float type

steppi committed:
TST: Fix reference values for some cdf stuff

steppi committed:
MAINT: Simplification and refactor for for discrete inversion

steppi committed:
MAINT: Vectorize pmf call in discrete mode calculation

steppi committed:
MAINT: Remove unnecessary edge case handling

steppi committed:
MAINT: Fix section that was mangled when resolving merge conflicts

steppi committed:
MAINT: Use ugly fix _cdf2_input_validation

steppi committed:
MAINT: Refector _cdf2_subtraction/safe

steppi committed:
MAINT: Change way of accessing one parameters array

steppi committed:
MAINT: Fix previous mistake resolving merge conflicts

mdhaber committed:
Merge pull request #121 from steppi/DiscreteDistributionSuggestions

Discrete distribution suggestions

mdhaber commented:
An MRE for one of the hypothesis failures is:
```python3
import numpy as np
from scipy import stats
x = np.asarray([np.nan, np.nan])
y = np.array([ 6., 10.])
X = stats.Binomial(n=np.array(15.), p=np.array([0.65663512, 0.25826382]))
X.logccdf(x, y, method='addition')
# array([-0.03805847, -8.74909714])
```
Also I saw that the return type of `logcdf` is still complex, but it can never by negative now.

mdhaber commented:
The refguide failure hints at a mistake in merge conflict resolution at some point. Some of the new `_make_distribution_custom` code didn't make it, apparently. We'll have to skim the diff here for accidental regressions. (Presumably there aren't many if tests are passing, though.)

mdhaber committed:
MAINT: stats.DiscreteDistribution: fix mistake in merge / accidental reversion

mdhaber committed:
STY: stats.DiscreteDistribution: fix PEP8 line length

steppi committed:
Revert "MAINT: Update cdf2 and friends to use probability definition"

This reverts commit 02819acf234f0e01c41c983594a52bcd6443239e.

steppi committed:
Revert "TST: Update tests in response to changes in cdf2 and friends"

This reverts commit e171a151fd4340ef04bcda36bc2e1fe00d9a2523.

steppi committed:
Revert "DOC: Update definition in docs for two arg ccdf"

This reverts commit 29adf2b8f59148ec8e38596a0825bd2bceb44c48.

steppi committed:
Revert "TST: Fix reference values for some cdf stuff"

This reverts commit 63d4adde7d5cd909f96c6594bf736fe7fc6e0432.

steppi committed:
MAINT: Make two argument cdf raise for discrete distributions

steppi committed:
TST: Test that two arg cdf functions for discrete distributions raise

mdhaber committed:
Merge pull request #122 from steppi/DiscreteDistributionSuggestionCleanup

Discrete distribution suggestion cleanup

mdhaber committed:
MAINT: stats.DiscreteDistribution: resolve merge conflict

mdhaber commented:
One failure is still:
```python3
from scipy import stats
X = stats.Binomial(n=18, p=1)
X.moment(3, kind='central', method='normalize')
```
Not raising on all platforms.

The one on the mypy job is:
```python3
import numpy as np
from scipy import stats
X = stats.Binomial(n=np.asarray([10., 16.]), p=np.asarray([0., 0.]))
X.iccdf(0.99156979), X.iccdf(0.99156979, method='inversion') 
# (array([0., 0.]), array([10., 16.]))
```

mdhaber committed:
STY: stats.DiscreteDistribution: fix PEP8

mdhaber committed:
MAINT: stats.Binomial: make p exclusive of endpoints

mdhaber committed:
MAINT: stats.Binomial: exclude n=0 degenerate distribution

steppi commented:
The failure is unrelated; I also see it here https://github.com/scipy/scipy/pull/22939. Today I learned that Sphinx uses the [Snowball Stemmer](https://pypi.org/project/snowballstemmer/) to reduce words to a root form for use in search. CI was unlucky and pulled a broken release of the snowball stemmer before it was yanked, see https://github.com/snowballstem/snowball/issues/229. This has since been resolved upstream.

steppi reviewed (approved):
I'm happy with the state this is in. There are some important things to follow-up on. Offhand,

1. Find and stick with a consistent two argument cdf definition for both continuous and discrete distributions. 
2. Get rid of hypothesis, and reorganize the test suite so that tests are determinstic, with failures in one assert not blocking remaining asserts from running, and skip logic can be done at the pytest mark level, not in conditionals spread throughout the code.
3. Decide on what to do about functions currently needing to be responsible for knowing when the output should be nan, and whether we really want to require nan checks strewn throughout various function implementations to make things work. We may want the NaN checks in the input validation instead.
4. A possible improvement @mdhaber had suggested to the mode calculation for discrete distributions, https://github.com/mdhaber/scipy/pull/121#discussion_r2059267607.

Very nice work @mdhaber. Sorry I couldn't get the probability definition for two argument cdf working, it seems it will require some thought. I'm going to regular merge this so that our separate work doesn't get blended together in the git blame.

mdhaber commented:
A few other items for follow-up:

- investigate/fix test skips (e.g. discrete distributions that don't work with `make_distribution`
- fix slow inversion / edge case issues when bracket search fails; this should fix degenerate `Binomial` distributions
- https://github.com/scipy/scipy/pull/22312#issuecomment-2588114565
